src/main/kotlin/uk/gov/justice/digital/hmpps/educationandworkplanapi/app/database/jpa/JpaInductionPersistenceAdapter.kt

```diff
@@ -46,4 +56,55 @@ class JpaInductionPersistenceAdapter(
     inductionRepository.findByPrisonNumberIn(prisonNumbers).let {
       inductionMapper.fromEntitySummariesToDomainSummaries(it)
     }
+
+  private fun createOrUpdatePreviousQualifications(updateInductionDto: UpdateInductionDto, prisonNumber: String): PreviousQualificationsEntity? {
+    val previousQualificationsEntity = previousQualificationsRepository.findByPrisonNumber(prisonNumber)
+
+    if (updateInductionDto.previousQualifications == null) {
+      // If previousQualifications on the DTO is null it means no change is required to the prisoner's previousQualifications. Return the entity.
+      return previousQualificationsEntity
+    }
+
+    // The DTO contains previousQualifications. Behaviour now depends on whether the prisoner already has previousQualifications or not.
+    if (previousQualificationsEntity != null) {
+      // Prisoner already has previousQualifications. We need to update them.
+      previousQualificationsMapper.updateExistingEntityFromDto(
+        previousQualificationsEntity,
+        updateInductionDto.previousQualifications,
+      )
+      return previousQualificationsRepository.saveAndFlush(previousQualificationsEntity)
+    } else {
+      // Prisoner does not already have previous qualifications, and the DTO has qualifications. We need to create them
+      return previousQualificationsRepository.saveAndFlush(
+        previousQualificationsMapper.fromUpdateDtoToNewEntity(updateInductionDto.previousQualifications)!!,
+      )
+    }
+  }
+
+  private fun createOrUpdatePreviousQualifications(createInductionDto: CreateInductionDto, prisonNumber: String): PreviousQualificationsEntity? {
+    val previousQualificationsEntity = previousQualificationsRepository.findByPrisonNumber(prisonNumber)
+
+    if (createInductionDto.previousQualifications == null) {
+      // If previousQualifications on the DTO is null it means the CIAG is creating an Induction with the specific intent of the prisoner having no previous qualifications.
+      // In this case we should delete the previousQualifications entity if it exists (because regardless of any previously added qualifications for the prisoner, the CIAG is
+      // saying that the Induction should be created with no qualifications)
+      previousQualificationsEntity?.also { previousQualificationsRepository.delete(it) }
```

Comment:
This makes me slightly nervous but I assume we'll have loaded the qualifications during the induction and the CIAG would've explicitly deleted them during induction?

Reply:
Yes, I agree, I'm equally nervous about it; but it does feel like the right outcome.

The scenario is that a prisoner has had qualifications added before induction
Then, at some point after, the CIAG does the induction. As part of doing the induction their previously entered quals are retrieved and pre-setup and displayed on the Qualifications page of the induction.

* If the CIAG is happy to accept them, they can move on through the induction, then at the end when we call the API it will be a case of the prisoner already has quals, and the DTO has quals, and they are the same
* If the CIAG wants to edit them (remove one or more, add some, etc) they can, and they move on through the induction. Then at the end when we call the API it will be a case of the prisoner already has quals, and the DTO has quals, and they are different so the quals in the database get updated
* If the CIAG is not happy with them all and removes them all, then this is the use case we are talking about here. In this case the existing quals for the prisoner should be deleted.

Personally I think it's a real edge case in respect of numbers of people actually doing it, but I think we do need to support it.
I think I'll add a log statement or two so that we can at least retrospectively see that people have done it.

Reply:
I've added some logging, which has highlighted there are 3 cases where you can end up with no qualifications
Of those 3 only one of them is the JPA `delete` that you've highlighted in this comment. 
The more I think about it the more I think it wont happen in practice (the other 2 will)
I think it's right to keep the code in because it _could_ happen, but only if we changed how our UI creates and sends the Create Induction message (so that it sends a null PreviousQualifications object - at the moment I dont think it ever would based on the question set changes we made a few weeks back)
But we might change that user journey (again 🤷‍♂️ ), or our API might get called by something other than our UI ? 🤷‍♂️ 

So I think it's right to have the code coverage for that eventuality